README.md
Live Profiler
================

![logo](images/liveprof_logo.png "logo")

Live Profiler is a system-wide performance monitoring system in use at Badoo that is built on top of [XHProf](http://pecl.php.net/package/xhprof) or its forks ([Uprofiler](https://github.com/FriendsOfPHP/uprofiler) or [Tideways](https://github.com/tideways/php-profiler-extension)).
Live Profiler continually gathers function-level profiler data from production tier by running a sample of page requests under XHProf.

[Live profiler UI](https://github.com/badoo/liveprof-ui) then aggregates the profile data corresponding to individual requests by various dimensions such a time, page type, and can help answer a variety of questions such as:
What is the function-level profile for a specific page?
How expensive is function "foo" across all pages, or on a specific page?
What functions regressed most in the last day/week/month?
What is the historical trend for execution time of a page/function? and so on.

Here is [a plugin](https://plugins.jetbrains.com/plugin/13767-live-profiler) for PhpStorm to see the method performance directly in IDE.

[liveprof.org](http://liveprof.org/) shows all features and can be used for test purposes.

[![Build Status](https://travis-ci.org/badoo/liveprof.svg?branch=master)](https://travis-ci.org/badoo/liveprof)
[![codecov](https://codecov.io/gh/badoo/liveprof/branch/master/graph/badge.svg)](https://codecov.io/gh/badoo/liveprof)
[![Scrutinizer Code Quality](https://scrutinizer-ci.com/g/badoo/liveprof/badges/quality-score.png?b=master)](https://scrutinizer-ci.com/g/badoo/liveprof/?branch=master)
[![GitHub license](https://img.shields.io/github/license/badoo/liveprof.svg)](https://github.com/badoo/liveprof/blob/master/LICENSE)

System Requirements
===================

* PHP version 5.4 or later / hhvm version 3.25.0 or later
* One of [XHProf](http://pecl.php.net/package/xhprof),
     [Uprofiler](https://github.com/FriendsOfPHP/uprofiler) or
     [Tideways](https://github.com/tideways/php-profiler-extension) to profile and collect the data.
  You can use other profiler which returns data in the follow format:
  ```php
  $data = [
      [
          'parent_method==>child_method' => [
              'param' => 'value' 
          ]
      ]  
  ];
  ```  
* Database extension to save profiling results to the database.

Installation
============

1. You can install Live Profiler via [Composer](https://getcomposer.org/):

```bash
php composer.phar require badoo/liveprof
```

2. Prepere a storage for results depends on mode

[save data in database] If you use DB mode you need to prepare a database server. You can use any driver described [here](https://www.doctrine-project.org/projects/doctrine-dbal/en/2.8/reference/configuration.html#configuration) or implement the custom one. You need run a script to configure database. This script creates "details" table: 

```bash
LIVE_PROFILER_CONNECTION_URL=mysql://db_user:db_password@db_mysql:3306/Profiler?charset=utf8 php vendor/badoo/liveprof/bin/install.php
```

[save data in files] It's also possible to save profiling result into files. To do it prepare a directory with write permissions.

[send data to demo site] You need to visit [liveprof.org](http://liveprof.org/) , sign in and copy API key.

3. Init a profiler before working code in the project entry point (usually public/index.php).

Usage
=====

There is an example of usage a profiler with default parameters:
```php
<?php
include 'vendor/autoload.php';

\Badoo\LiveProfiler\LiveProfiler::getInstance()->start();
// Code is here
```

There is an example how to test Live Profiler without any extension and database. You can use a build-in profiler compatible with XHProf and <a href="http://liveprof.org">liveprof.org</a> as UI:
```php
<?php
include 'vendor/autoload.php';

\Badoo\LiveProfiler\LiveProfiler::getInstance()
     ->setMode(\Badoo\LiveProfiler\LiveProfiler::MODE_API)
     ->setApiKey('70366397-97d6-41be-a83c-e9e649c824e1') // a key for guest
     ->useSimpleProfiler() // Use build-in profiler instead of XHProf or its forks
     ->setApp('Demo') // Some unique app name
     ->start();
     
// Code is here
// start a timer before each inportant method
\Badoo\LiveProfiler\SimpleProfiler::getInstance()->startTimer(__METHOD__); // any string can be used as a timer tag
// stop the timer before the end of the method
\Badoo\LiveProfiler\SimpleProfiler::getInstance()->endTimer(__METHOD__); // any string can be used as a timer tag
```

There is a full list of methods you can use to change options:
```php
<?php

// Start profiling
\Badoo\LiveProfiler\LiveProfiler::getInstance()
    ->setMode(\Badoo\LiveProfiler\LiveProfiler::MODE_DB) // optional, MODE_DB - save profiles to db, MODE_FILES - save profiles to files, MODE_API - send profiles to http://liveprof.org/ 
    ->setConnectionString('mysql://db_user:db_password@db_mysql:3306/Profiler?charset=utf8') // optional, you can also set the connection url in the environment variable LIVE_PROFILER_CONNECTION_URL
    ->setPath('/app/data/') // optional, path to save profiles, you can also set the file path in the environment variable LIVE_PROFILER_PATH
    ->setApiKey('api_key') // optional, api key to send profiles and see demo, you can get it on http://liveprof.org/ 
    ->setApp('Site1') // optional, current app name to use one profiler in several apps, "Default" by default
    ->setLabel('users') // optional, the request name, by default the url path or script name in cli
    ->setDivider(700) // optional, profiling starts for 1 of 700 requests with the same app and label, 1000 by default
    ->setTotalDivider(7000) // optional, profiling starts for 1 of 7000 requests with forces label "All", 10000 by default
    ->setLogger($Logger) // optional, a custom logger implemented \Psr\Log\LoggerInterface
    ->setConnection($Connection) // optional, a custom instance of \Doctrine\DBAL\Connection if you can't use the connection url
    ->setDataPacker($DatePacker) // optional, a class implemented \Badoo\LiveProfiler\DataPackerInterface to convert array into string
    ->setStartCallback($profiler_start_callback) // optional, set it if you use custom profiler
    ->setEndCallback($profiler_profiler_callback) // optional, set it if you use custom profiler
    ->useXhprof() // optional, force use xhprof as profiler
    ->useTidyWays() // optional, force use TidyWays as profiler
    ->useUprofiler() // optional, force use uprofiler as profiler
    ->useSimpleProfiler() // optional, force use internal profiler
    ->useXhprofSample() // optional, force use xhprof in sampling mode
    ->start();
```

If you want to change the Label during running (for instance, after you got some information in the router or controller) you can call:
```php
<?php

$number = random_int(0, 100);
$current_label = \Badoo\LiveProfiler\LiveProfiler::getInstance()->getLabel();
\Badoo\LiveProfiler\LiveProfiler::getInstance()->setLabel($current_label . $number);
```

if you don't want to save profiling result you can reset it anytime:
```php
<?php

\Badoo\LiveProfiler\LiveProfiler::getInstance()->reset();
```

After script ends it will call `\Badoo\LiveProfiler\LiveProfiler::getInstance()->end();` on shutdown, but you can call it explicitly after working code.

Environment Variables
=====================

`LIVE_PROFILER_CONNECTION_URL`: [url](https://www.doctrine-project.org/projects/doctrine-dbal/en/2.8/reference/configuration.html#configuration) for the database connection

`LIVE_PROFILER_PATH`: path to save profiles in \Badoo\LiveProfiler\LiveProfiler::MODE_FILES mode

`LIVE_PROFILER_API_URL`: api url to send profiles in \Badoo\LiveProfiler\LiveProfiler::MODE_API mode and see demo on [liveprof.org](http://liveprof.org/) 

Work flow
=========

Live profiler allows to run profiling with custom frequency (for instance 1 of 1000 requests) grouped by app name ('Default' by default) and custom label (by default it's the url path or script name).

It's important to calculate **the request divider** properly to have enough data for aggregation. You should divide daily request count to have approximately 1 profile per minute.
For instance, if you have 1M requests a day for the page /users the divider should be 1000000/(60*24) = 694, so divider = 700 is enough.

Also you have to calculate **a total divider** for all request profiling. It's important to control the whole system health. It can be calculated the same way as a divider calculation for particularly request,
but in this case you should use count of all daily requests. For instance, if you have 10M requests a day - total_divider=10000000/(60*24) = 6940, so total_divider = 7000 is enough.

The profiler automatically detects which profiler extension you have (xhprof, uprofiler or tidyways). You have to set profiler callbacks if you use other profiler.

You can run the test script in the docker container with **xhprof** extension and look at the example of the server configuration in Dockerfile:
```bash 
docker build -t badoo/liveprof .
docker run badoo/liveprof
```

or you can build a docker container with **xhprof** using sampling:
```bash 
docker build -f DockerfileSamples -t badoo/liveprof .
docker run badoo/liveprof
```

or you can build a docker container with **tideways** extension:
```bash 
docker build -f DockerfileTidyWays -t badoo/liveprof .
docker run badoo/liveprof
```

or **uprofiler** extension:
```bash 
docker build -f DockerfileUprofiler -t badoo/liveprof .
docker run badoo/liveprof
```

or latest hhvm with included  **xhprof** extension:
```bash
docker build -f DockerfileHHVM -t badoo/liveprof .
docker run badoo/liveprof
```
or if you want to use API with included  **xhprof** extension:
```bash
docker build -f DockerfileUseApi -t badoo/liveprof .
docker run badoo/liveprof
```

If your server has php version 7.0 or later it's better to use [Tideways](https://github.com/tideways/php-profiler-extension) as profiler.

Steps to install tideways extension:
```bash
git clone https://github.com/tideways/php-profiler-extension.git
cd php-profiler-extension
phpize
./configure
make
make install
echo "extension=tideways_xhprof.so" >> /usr/local/etc/php/conf.d/20-tideways_xhprof.ini
echo "xhprof.output_dir='/tmp/xhprof'" >> /usr/local/etc/php/conf.d/20-tideways_xhprof.ini
```

Steps to install uprofiler:
```bash
git clone https://github.com/FriendsOfPHP/uprofiler.git
cd uprofiler/extension/
phpize
./configure
make
make install
echo "extension=uprofiler.so" >> /usr/local/etc/php/conf.d/20-uprofiler.ini
echo "uprofiler.output_dir='/tmp/uprofiler'" >> /usr/local/etc/php/conf.d/20-uprofiler.ini
```

Tests
=====

Install Live Profiler with dev requirements:
```bash 
php composer.phar require --dev badoo/liveprof
```

In the project directory, run:
```bash
vendor/bin/phpunit
```

License
=======

This project is licensed under the MIT open source license.


CHANGELOG.md
# Live Profiler ChangeLog

## master

There are next changes:

## 1.3.4

There are next changes:

- added a memory  usage for SimpleProfiler
- bug fixes

## 1.3.3

There are next changes:

- added SimpleProfiler to use liveprof without any extension

## 1.3.2

There are next changes:

- changed version of doctrine/dbal
- added php 7.3 support

## 1.3.1

There are next changes:

- added example of sending profiling results to api

## 1.3.0

There are next changes:

- added sending profile to api

## 1.2.0

There are next changes:

- added xhprof_sample support

## 1.1.0

There are next changes:

- added possibility to save profiles in the files


tests/bootstrap.php
<?php
// for backward compatible


tests/unit/Badoo/BaseTestCase.php
<?php

/**
 * @maintainer Timur Shagiakhmetov <timur.shagiakhmetov@corp.badoo.com>
 */

namespace unit\Badoo;

class BaseTestCase extends \PHPUnit\Framework\TestCase
{
    /**
     * Call protected/private method of a class.
     * @param object &$object    Instantiated object that we will run method on.
     * @param string $methodName Method name to call
     * @param array  $parameters Array of parameters to pass into method.
     * @return mixed Method return.
     * @throws \ReflectionException
     */
    public function invokeMethod(&$object, $methodName, array $parameters = array())
    {
        $reflection = new \ReflectionClass(\get_class($object));
        $method = $reflection->getMethod($methodName);
        $method->setAccessible(true);

        return $method->invokeArgs($object, $parameters);
    }

    /**
     * @param $object
     * @param $property
     * @param $value
     * @throws \ReflectionException
     */
    public function setProtectedProperty(&$object, $property, $value)
    {
        $reflection = new \ReflectionClass(\get_class($object));
        $reflection_property = $reflection->getProperty($property);
        $reflection_property->setAccessible(true);
        $reflection_property->setValue($object, $value);
    }

    /**
     * @param $object
     * @param $property
     * @return mixed
     * @throws \ReflectionException
     */
    public function getProtectedProperty(&$object, $property)
    {
        $reflection = new \ReflectionClass(\get_class($object));
        $reflection_property = $reflection->getProperty($property);
        $reflection_property->setAccessible(true);
        return $reflection_property->getValue($object);
    }
}


tests/unit/Badoo/LiveProfiler/LoggerTest.php
<?php

/**
 * @maintainer Timur Shagiakhmetov <timur.shagiakhmetov@corp.badoo.com>
 */

namespace unit\Badoo\LiveProfiler;

class LoggerTest extends \unit\Badoo\BaseTestCase
{
    /**
     * @throws \ReflectionException
     */
    public function testGetLogMsg()
    {
        $Logger = new \Badoo\LiveProfiler\Logger();
        $log_msg = $this->invokeMethod($Logger, 'getLogMsg', ['error', 'Error msg', ['param' => 1]]);
        self::assertEquals(date('Y-m-d H:i:s'). "\terror\tError msg\t{\"param\":1}\n", $log_msg);
    }

    public function testLog()
    {
        $tmp_log_file = tempnam('/tmp', 'live.profiling');
        $Logger = new \Badoo\LiveProfiler\Logger();
        $Logger->setLogFile($tmp_log_file);
        $Logger->log('error', 'Error msg');

        $log_msg = file_get_contents($tmp_log_file);
        unset($tmp_log_file);

        self::assertEquals(date('Y-m-d H:i:s'). "\terror\tError msg\n", $log_msg);
    }
}


src/Badoo/LiveProfiler/Logger.php
<?php

/**
 * @maintainer Timur Shagiakhmetov <timur.shagiakhmetov@corp.badoo.com>
 */

namespace Badoo\LiveProfiler;

use Psr\Log\LoggerInterface;
use Psr\Log\LoggerTrait;

class Logger implements LoggerInterface
{
    use LoggerTrait;

    /** @var string */
    protected $logfile;

    /**
     * Logger constructor.
     */
    public function __construct()
    {
        $this->logfile = __DIR__ . '/../../../live.profiler.log';
    }

    public function setLogFile($logfile)
    {
        $this->logfile = $logfile;
    }

    /**
     * @param mixed $level
     * @param string $message
     * @param array $context
     */
    public function log($level, $message, array $context = array())
    {
        $log_string = $this->getLogMsg($level, $message, $context);
        file_put_contents($this->logfile, $log_string, FILE_APPEND);
    }

    /**
     * @param string $level
     * @param string $message
     * @param array $context
     * @return string
     */
    protected function getLogMsg($level, $message, array $context = array())
    {
        $log_string = sprintf("%s\t%s\t%s", date('Y-m-d H:i:s'), $level, $message);

        if (!empty($context)) {
            $log_string .= "\t" . json_encode($context, true);
        }

        $log_string .= "\n";

        return $log_string;
    }
}


src/Badoo/LiveProfiler/SimpleProfiler.php
<?php

/**
 * @maintainer Timur Shagiakhmetov <timur.shagiakhmetov@corp.badoo.com>
 */

namespace Badoo\LiveProfiler;

class SimpleProfiler
{
    /**
     * @var self
     */
    private static $instance;

    private $methods = [];
    /** @var \SplStack */
    private $stack;
    private $is_enabled = false;

    public static function getInstance()
    {
        if (self::$instance === null) {
            self::$instance = new static();
        }

        return self::$instance;
    }

    public function enable()
    {
        $this->is_enabled= true;
        $this->stack = new \SplStack();

        $this->startTimer('main()');
    }

    public function disable()
    {
        $this->endTimer('main()');

        if (!$this->stack->isEmpty()) {
            return [];
        }

        $this->is_enabled = false;

        return $this->methods;
    }

    public function startTimer($tag)
    {
        if (!$this->is_enabled) {
            return false;
        }

        $full_method = $this->getFullMethod($tag);
        $this->stack->push($tag);

        if (!isset($this->methods[$full_method])) {
            $this->methods[$full_method] = [
                'wt' => 0,
                'mu' => 0,
                'ct' => 0,
            ];
        }

        $time = (int)(microtime(true) * 1e6);
        $memory = memory_get_usage();
        $this->methods[$full_method]['wt'] = $time - $this->methods[$full_method]['wt'];
        $this->methods[$full_method]['mu'] = $memory - $this->methods[$full_method]['mu'];

        return true;
    }

    public function endTimer($tag)
    {
        if (!$this->is_enabled) {
            return false;
        }

        if ($this->stack->isEmpty() || $this->stack->top() !== $tag) {
            return false;
        }

        $tag = $this->stack->pop();

        $full_method = $this->getFullMethod($tag);

        $time = (int)(microtime(true) * 1e6);
        $memory = memory_get_usage();
        $this->methods[$full_method]['ct']++;
        $this->methods[$full_method]['wt'] = $time - $this->methods[$full_method]['wt'];
        $this->methods[$full_method]['mu'] = $memory - $this->methods[$full_method]['mu'];

        return true;
    }

    private function getFullMethod($tag)
    {
        if ($this->stack->isEmpty()) {
            return $tag;
        }

        return $this->stack->top() . '==>' . $tag;
    }
}


src/Badoo/LiveProfiler/LiveProfiler.php
<?php

/**
 * @maintainer Timur Shagiakhmetov <timur.shagiakhmetov@corp.badoo.com>
 */

namespace Badoo\LiveProfiler;

use Doctrine\DBAL\Connection;
use Doctrine\DBAL\DBALException;
use Psr\Log\LoggerInterface;

class LiveProfiler
{
    CONST MODE_DB = 'db';
    CONST MODE_FILES = 'files';
    CONST MODE_API = 'api';

    /** @var LiveProfiler */
    protected static $instance;
    /** @var string */
    protected $mode = self::MODE_DB;
    /** @var string */
    protected $path = '';
    /** @var string */
    protected $api_key = '';
    /** @var string */
    protected $url = 'http://liveprof.org/api';
    /** @var Connection */
    protected $Conn;
    /** @var LoggerInterface */
    protected $Logger;
    /** @var DataPackerInterface */
    protected $DataPacker;
    /** @var string */
    protected $connection_string;
    /** @var string */
    protected $app;
    /** @var string */
    protected $label;
    /** @var string */
    protected $datetime;
    /** @var int */
    protected $divider = 1000;
    /** @var int */
    protected $total_divider = 10000;
    /** @var callable callback to start profiling */
    protected $start_callback;
    /** @var callable callback to end profiling */
    protected $end_callback;
    /** @var bool */
    protected $is_enabled = false;
    /** @var array */
    protected $last_profile_data = [];

    /**
     * LiveProfiler constructor.
     * @param string $connection_string_or_path
     * @param string $mode
     */
    public function __construct($connection_string_or_path = '', $mode = self::MODE_DB)
    {
        $this->mode = $mode;

        $this->app = 'Default';
        $this->label = $this->getAutoLabel();
        $this->datetime = date('Y-m-d H:i:s');

        $this->detectProfiler();
        $this->Logger = new Logger();
        $this->DataPacker = new DataPacker();

        if ($mode === self::MODE_DB) {
            $this->connection_string = $connection_string_or_path ?: getenv('LIVE_PROFILER_CONNECTION_URL');
        } elseif ($mode === self::MODE_API) {
            if ($connection_string_or_path) {
                $this->url = $connection_string_or_path;
            } elseif (getenv('LIVE_PROFILER_API_URL')) {
                $this->url = getenv('LIVE_PROFILER_API_URL');
            }
        } else {
            $this->setPath($connection_string_or_path ?: getenv('LIVE_PROFILER_PATH'));
        }
    }

    public static function getInstance($connection_string = '', $mode = self::MODE_DB)
    {
        if (self::$instance === null) {
            self::$instance = new static($connection_string, $mode);
        }

        return self::$instance;
    }

    public function start()
    {
        if ($this->is_enabled) {
            return true;
        }

        if (null === $this->start_callback) {
            return true;
        }

        if ($this->needToStart($this->divider)) {
            $this->is_enabled = true;
        } elseif ($this->needToStart($this->total_divider)) {
            $this->is_enabled = true;
            $this->label = 'All';
        }

        if ($this->is_enabled) {
            register_shutdown_function([$this, 'end']);
            call_user_func($this->start_callback);
        }

        return true;
    }

    /**
     * @return bool
     */
    public function end()
    {
        if (!$this->is_enabled) {
            return true;
        }

        $this->is_enabled = false;

        if (null === $this->end_callback) {
            return true;
        }

        $data = call_user_func($this->end_callback);
        if (!is_array($data)) {
            $this->Logger->warning('Invalid profiler data: ' . var_export($data, true));
            return false;
        }

        if (empty($data)) {
            return false;
        }

        $this->last_profile_data = $data;
        $result = $this->save($this->app, $this->label, $this->datetime, $data);

        if (!$result) {
            $this->Logger->warning('Can\'t insert profile data');
        }

        return $result;
    }

    public function detectProfiler()
    {
        if (function_exists('xhprof_enable')) {
            return $this->useXhprof();
        }

        if (function_exists('tideways_xhprof_enable')) {
            return $this->useTidyWays();
        }

        if (function_exists('uprofiler_enable')) {
            return $this->useUprofiler();
        }

        return $this->useSimpleProfiler();
    }

    public function useXhprof()
    {
        if ($this->is_enabled) {
            $this->Logger->warning('can\'t change profiler after profiling started');
            return $this;
        }

        $this->start_callback = function () {
            xhprof_enable(XHPROF_FLAGS_MEMORY | XHPROF_FLAGS_CPU);
        };

        $this->end_callback = function () {
            return xhprof_disable();
        };

        return $this;
    }

    public function useXhprofSample()
    {
        if ($this->is_enabled) {
            $this->Logger->warning('can\'t change profiler after profiling started');
            return $this;
        }

        if (!ini_get('xhprof.sampling_interval')) {
            ini_set('xhprof.sampling_interval', 10000);
        }

        if (!ini_get('xhprof.sampling_depth')) {
            ini_set('xhprof.sampling_depth', 200);
        }

        $this->start_callback = function () {
            define('XHPROF_SAMPLING_BEGIN', microtime(true));
            xhprof_sample_enable();
        };

        $this->end_callback = function () {
            return $this->convertSampleDataToCommonFormat(xhprof_sample_disable());
        };

        return $this;
    }

    protected function convertSampleDataToCommonFormat(array $sampling_data)
    {
        $result_data = [];
        $prev_time = XHPROF_SAMPLING_BEGIN;
        foreach ($sampling_data as $time => $callstack) {
            $wt = (int)(($time - $prev_time) * 1000000);
            $functions = explode('==>', $callstack);
            $prev_i = 0;
            $main_key = $functions[$prev_i];
            if (!isset($result_data[$main_key])) {
                $result_data[$main_key] = [
                    'ct' => 0,
                    'wt' => 0,
                ];
            }
            $result_data[$main_key]['ct'] ++;
            $result_data[$main_key]['wt'] += $wt;

            $func_cnt = count($functions);
            for ($i = 1; $i < $func_cnt; $i++) {
                $key = $functions[$prev_i] . '==>' . $functions[$i];

                if (!isset($result_data[$key])) {
                    $result_data[$key] = [
                        'ct' => 0,
                        'wt' => 0,
                    ];
                }

                $result_data[$key]['wt'] += $wt;
                $result_data[$key]['ct']++;

                $prev_i = $i;
            }

            $prev_time = $time;
        }

        return $result_data;
    }

    public function useTidyWays()
    {
        if ($this->is_enabled) {
            $this->Logger->warning('can\'t change profiler after profiling started');
            return $this;
        }

        $this->start_callback = function () {
            tideways_xhprof_enable(TIDEWAYS_XHPROF_FLAGS_MEMORY | TIDEWAYS_XHPROF_FLAGS_CPU);
        };

        $this->end_callback = function () {
            return tideways_xhprof_disable();
        };

        return $this;
    }

    public function useUprofiler()
    {
        if ($this->is_enabled) {
            $this->Logger->warning('can\'t change profiler after profiling started');
            return $this;
        }

        $this->start_callback = function () {
            uprofiler_enable(UPROFILER_FLAGS_CPU | UPROFILER_FLAGS_MEMORY);
        };

        $this->end_callback = function () {
            return uprofiler_disable();
        };

        return $this;
    }

    public function useSimpleProfiler()
    {
        if ($this->is_enabled) {
            $this->Logger->warning('can\'t change profiler after profiling started');
            return $this;
        }

        $this->start_callback = function () {
            \Badoo\LiveProfiler\SimpleProfiler::getInstance()->enable();
        };

        $this->end_callback = function () {
            return \Badoo\LiveProfiler\SimpleProfiler::getInstance()->disable();
        };

        return $this;
    }

    /**
     * @return bool
     */
    public function reset()
    {
        if ($this->is_enabled) {
            call_user_func($this->end_callback);
            $this->is_enabled = false;
        }

        return true;
    }

    /**
     * @param string $mode
     * @return $this
     */
    public function setMode($mode)
    {
        $this->mode = $mode;
        return $this;
    }

    /**
     * @return string
     */
    public function getMode()
    {
        return $this->mode;
    }

    /**
     * @param string $path
     * @return $this
     */
    public function setPath($path)
    {
        if (!is_dir($path)) {
            $this->Logger->error('Directory ' . $path . ' does not exists');
        }

        $this->path = $path;
        return $this;
    }

    /**
     * @return string
     */
    public function getPath()
    {
        return $this->path;
    }

    /**
     * @param string $api_key
     * @return $this
     */
    public function setApiKey($api_key)
    {
        $this->api_key = $api_key;
        return $this;
    }

    /**
     * @return string
     */
    public function getApiKey()
    {
        return $this->api_key;
    }

    /**
     * @param string $app
     * @return $this
     */
    public function setApp($app)
    {
        $this->app = $app;
        return $this;
    }

    /**
     * @return string
     */
    public function getApp()
    {
        return $this->app;
    }

    /**
     * @param string $label
     * @return $this
     */
    public function setLabel($label)
    {
        $this->label = $label;
        return $this;
    }

    /**
     * @return string
     */
    public function getLabel()
    {
        return $this->label;
    }

    /**
     * @param string $datetime
     * @return $this
     */
    public function setDateTime($datetime)
    {
        $this->datetime = $datetime;
        return $this;
    }

    /**
     * @return string
     */
    public function getDateTime()
    {
        return $this->datetime;
    }

    /**
     * @param int $divider
     * @return $this
     */
    public function setDivider($divider)
    {
        $this->divider = $divider;
        return $this;
    }

    /**
     * @param int $total_divider
     * @return $this
     */
    public function setTotalDivider($total_divider)
    {
        $this->total_divider = $total_divider;
        return $this;
    }

    /**
     * @param \Closure $start_callback
     * @return $this
     */
    public function setStartCallback(\Closure $start_callback)
    {
        $this->start_callback = $start_callback;
        return $this;
    }

    /**
     * @param \Closure $end_callback
     * @return $this
     */
    public function setEndCallback(\Closure $end_callback)
    {
        $this->end_callback = $end_callback;
        return $this;
    }

    /**
     * @param LoggerInterface $Logger
     * @return $this
     */
    public function setLogger(LoggerInterface $Logger)
    {
        $this->Logger = $Logger;
        return $this;
    }

    /**
     * @param DataPackerInterface $DataPacker
     * @return $this
     */
    public function setDataPacker($DataPacker)
    {
        $this->DataPacker = $DataPacker;
        return $this;
    }

    /**
     * @return array
     */
    public function getLastProfileData()
    {
        return $this->last_profile_data;
    }

    /**
     * @return Connection
     * @throws DBALException
     */
    protected function getConnection()
    {
        if (null === $this->Conn) {
            $config = new \Doctrine\DBAL\Configuration();
            $connectionParams = ['url' => $this->connection_string];
            $this->Conn = \Doctrine\DBAL\DriverManager::getConnection($connectionParams, $config);
        }

        return $this->Conn;
    }

    /**
     * @param Connection $Conn
     * @return $this
     */
    public function setConnection(Connection $Conn)
    {
        $this->Conn = $Conn;
        return $this;
    }

    /**
     * @param string $connection_string
     * @return $this
     */
    public function setConnectionString($connection_string)
    {
        $this->connection_string = $connection_string;
        return $this;
    }

    /**
     * @param string $app
     * @param string $label
     * @param string $datetime
     * @param array $data
     * @return bool
     */
    protected function save($app, $label, $datetime, $data)
    {
        if ($this->mode === self::MODE_DB) {
            return $this->saveToDB($app, $label, $datetime, $data);
        }

        if ($this->mode === self::MODE_API) {
            return $this->sendToAPI($app, $label, $datetime, $data);
        }

        return $this->saveToFile($app, $label, $datetime, $data);
    }

    /**
     * @param string $app
     * @param string $label
     * @param string $datetime
     * @param array $data
     * @return bool
     */
    protected function sendToAPI($app, $label, $datetime, $data)
    {
        $data = $this->DataPacker->pack($data);
        $api_key = $this->api_key;
        $curl_handle = curl_init();
        curl_setopt($curl_handle,CURLOPT_URL,$this->url);
        curl_setopt($curl_handle,CURLOPT_RETURNTRANSFER, true);
        curl_setopt($curl_handle, CURLOPT_POST, 1);
        curl_setopt($curl_handle, CURLOPT_POSTFIELDS, http_build_query(compact('api_key', 'app', 'label', 'datetime', 'data')));
        curl_exec($curl_handle);
        $http_code = curl_getinfo($curl_handle, CURLINFO_HTTP_CODE);
        curl_close($curl_handle);

        return $http_code === 200;
    }

    /**
     * @param string $app
     * @param string $label
     * @param string $datetime
     * @param array $data
     * @return bool
     */
    protected function saveToDB($app, $label, $datetime, $data)
    {
        $packed_data = $this->DataPacker->pack($data);

        try {
            return (bool)$this->getConnection()->insert(
                'details',
                [
                    'app' => $app,
                    'label' => $label,
                    'perfdata' => $packed_data,
                    'timestamp' => $datetime
                ]
            );
        } catch (DBALException $Ex) {
            $this->Logger->error('Error in insertion profile data: ' . $Ex->getMessage());
            return false;
        }
    }

    /**
     * @param string $app
     * @param string $label
     * @param string $datetime
     * @param array $data
     * @return bool
     */
    private function saveToFile($app, $label, $datetime, $data)
    {
        $path = sprintf('%s/%s/%s', $this->path, $app, base64_encode($label));

        if (!is_dir($path) && !mkdir($path, 0755, true) && !is_dir($path)) {
            $this->Logger->error('Directory "'. $path .'" was not created');
            return false;
        }

        $filename = sprintf('%s/%s.json', $path, strtotime($datetime));
        $packed_data = $this->DataPacker->pack($data);
        return (bool)file_put_contents($filename, $packed_data);
    }

    /**
     * @throws DBALException
     */
    public function createTable()
    {
        $driver_name = $this->getConnection()->getDriver()->getName();
        $sql_path = __DIR__ . '/../../../bin/install_data/' . $driver_name . '/source.sql';
        if (!file_exists($sql_path)) {
            $this->Logger->error('Invalid sql path:' . $sql_path);
            return false;
        }

        $sql = file_get_contents($sql_path);

        $this->getConnection()->exec($sql);
        return true;
    }

    /**
     * @return string
     */
    protected function getAutoLabel()
    {
        if (!empty($_SERVER['REQUEST_URI'])) {
            $label = parse_url($_SERVER['REQUEST_URI'], PHP_URL_PATH);
            return $label ?: $_SERVER['REQUEST_URI'];
        }

        return $_SERVER['SCRIPT_NAME'];
    }

    /**
     * @param int $divider
     * @return bool
     */
    protected function needToStart($divider)
    {
        return mt_rand(1, $divider) === 1;
    }
}
